src/cdomains/apron/sharedFunctions.apron.ml

```diff
@@ -0,0 +1,300 @@
+        let texpr1_2 = texpr1_expr_of_cil_exp d env no_ov e2 in
+        (* Apron constraints always compare with 0 and only have comparisons one way *)
+        begin match r with
+          | Lt -> (texpr1_2, texpr1_1, SUP)   (* e1 < e2   ==>  e2 - e1 > 0  *)
+          | Gt -> (texpr1_1, texpr1_2, SUP)   (* e1 > e2   ==>  e1 - e2 > 0  *)
+          | Le -> (texpr1_2, texpr1_1, SUPEQ) (* e1 <= e2  ==>  e2 - e1 >= 0 *)
+          | Ge -> (texpr1_1, texpr1_2, SUPEQ) (* e1 >= e2  ==>  e1 - e2 >= 0 *)
+          | Eq -> (texpr1_1, texpr1_2, EQ)    (* e1 == e2  ==>  e1 - e2 == 0 *)
+          | Ne -> (texpr1_1, texpr1_2, DISEQ) (* e1 != e2  ==>  e1 - e2 != 0 *)
+          | _ -> raise Unsupported_CilExp
+        end
+      | _ -> raise Unsupported_CilExp
+    in
+    let inverse_typ = function
+      | EQ -> DISEQ
+      | DISEQ -> EQ
+      | SUPEQ -> SUP
+      | SUP -> SUPEQ
+      | EQMOD _ -> failwith "tcons1_of_cil_exp: cannot invert EQMOD"
+    in
+    let (texpr1_plus, texpr1_minus, typ) =
+      if negate then
+        (texpr1_minus, texpr1_plus, inverse_typ typ)
+      else
+        (texpr1_plus, texpr1_minus, typ)
+    in
+    let texpr1' = Binop (Sub, texpr1_plus, texpr1_minus, Int, Near) in
+    make (of_expr env texpr1') typ
+
+end
+
+module EnvOps =
+struct
+  let vars env =
+    let ivs, fvs = Environment.vars env in
+    assert (Array.length fvs = 0); (* shouldn't ever contain floats *)
+    List.of_enum (Array.enum ivs)
+
+  let mem_var env v = Environment.mem_var env v
+
+  let add_vars_with env vs =
+    let vs' =
+      vs
+      |> List.enum
+      |> Enum.filter (fun v -> not (Environment.mem_var env v))
+      |> Array.of_enum
+    in
+    Environment.add env vs' [||]
+
+  let remove_vars_with env vs =
+    let vs' =
+      vs
+      |> List.enum
+      |> Enum.filter (fun v -> Environment.mem_var env v)
+      |> Array.of_enum
+    in
+    Environment.remove env vs'
+
+  let get_filtered_vars_add env vs =
+    vs
+    |> List.enum
+    |> Enum.filter (fun v -> not (Environment.mem_var env v))
+    |> Array.of_enum
+
+  let get_filtered_vars_remove env vs =
+    vs
+    |> List.enum
+    |> Enum.filter (fun v -> Environment.mem_var env v)
+    |> Array.of_enum
+
+  let remove_filter_with env f =
+    let vs' =
+      vars env
+      |> List.enum
+      |> Enum.filter f
+      |> Array.of_enum
+    in
+    Environment.remove env vs'
+
+  let filter_vars env f =
+    vars env
+    |> List.enum
+    |> Enum.filter f
+    |> Array.of_enum
+
+  let keep_vars_with env vs =
+    let vs' =
+      vs
+      |> List.enum
+      |> Enum.filter (fun v -> Environment.mem_var env v)
+      |> Array.of_enum
+    in
+    Environment.make vs' [||]
+
+  let keep_filter_with env f =
+    (* Instead of removing undesired vars,
+       make a new env with just the desired vars. *)
+    let vs' =
+      vars env
+      |> List.enum
+      |> Enum.filter f
+      |> Array.of_enum
+    in
+    Environment.make vs' [||]
```

Comment:
Why are these `*_with` functions no longer in-place like they used to be? They instead copy the whole environment every time like the non-with ones used to.

Reply:
That was a naming issue, the `_with` ones also made use of these functions. The reason why those functions were put into `sharedFunctions` was that they were also used by `affineEqualityDomain` at a certain point. However, in the current state it just uses about 2 of them anymore. In the latest commit, I therefore moved some functions back again into `apronDomain` and put them in place again.